### PR TITLE
✨ Themes API activation permissions & validation

### PR DESCRIPTION
**Pull request by ErisDS**
Putting the permissions commit here + will need to do more work on how settings are called to make this not do permissions twice.

This change requires a db reset / fudging in order to get it to work.

UPDATE:

This PR now has validations & permissions for the activate endpoint. The setting is updated via the settings model not the edit endpoint, by passing permissions. I've removed active theme validation from the setting model - now the value will be validated, but nothing else.

- [x] ~~TODO: prevent setting active theme via /settings/~~
- [x] ~~TODO: tests~~
- [x] ~~TODO: make activate return single theme with warnings?~~

- [ ] TODO: make Ghost-Admin handle validation errors & warnings on activation, see https://github.com/TryGhost/Ghost/issues/8127

-----

refs #8093

Main changes:

- activation endpoint does permissions and validation
- activation endpoint returns warnings
- deleting active theme is not allowed
- changing active theme via settings is not allowed
- uploads & activations correctly return warnings
- much more, and better tests
- improved toJSON function replaces toAPI

### CODE REVIEW

**Comment by ErisDS** [comment]
Hey @kevinansfield I've been working on completing the themes API. There are a couple of notable updates here:

- activate endpoint now has proper permissions and validation
- delete endpoint will not allow you to delete the active theme (covered by the UI already)
- settings endpoint will not allow you to edit the activeTheme property

Whilst testing, I realised that Ghost-Admin doesn't handle ThemeValidationErrors. This got me thinking.

Should the activation endpoint return the full list as it does now? Or should it return a single theme more similarly to what the upload endpoint does.
The reason I'm asking is because, the upload endpoint returns a single object with warnings if they exist. The activation endpoint will not, currently return warnings.

It seems like it might be more consistent to return warnings from the activate endpoint?

Means the UI flow might get a little complicated. 

E.g. I upload a theme, it shows warnings. I click yes to activate, don't want to see warnings as I've already seen them.
E.g. I have a list of themes, I choose to activate one from the activate button, I should see warnings at this point?

I can add an extra commit here to change from full list to single + warnings if you agree 😬

Otherwise, this works is pretty much done. Cut off point for this PR is "Feature complete activation API".

**Comment by ErisDS** [comment]
I have updated this PR, hopefully this is a complete theme API for 1.0 for now 🎉 
Note: there is much more cleanup todo, I've only done a basic pass for now 

Main changes:

- activation endpoint does permissions and validation
- activation endpoint returns warnings
- deleting active theme is not allowed
- changing active theme via settings is not allowed
- uploads & activations correctly return warnings
- much more, and better tests
- improved toJSON function replaces toAPI

File: core/server/themes/validate.js [hunk 1]
```diff
@@ -1,21 +1,31 @@
 var Promise = require('bluebird'),
+    gscan = require('gscan'),
```
Comment (kirrg001) [inline core/server/themes/validate.js]: See https://github.com/TryGhost/Ghost/pull/8121/files#diff-8d96493a641ecbe40ffd7f378b4a175fL6
Would be great to not require gscan on bootstrap. Eats too much un-needed resources.

File: core/server/api/settings.js [hunk 2]
```diff
@@ -103,6 +103,15 @@ canEditAllSettings = function (settingsInfo, options) {
                 ));
             }
 
+            if (setting.key === 'activeTheme') {
```
Comment (kirrg001) [inline core/server/api/settings.js]: Hmm how about 
`if (setting.type === 'core' && !(options.context && options.context.internal)) {` replaced by
 `if ((setting.type === 'core' || setting.type === 'theme') && !(options.context && options.context.internal)) {`?

Or similar e.g. we make `activeTheme` part of `core`.
The goal is to have less extra checks here.

Comment (ErisDS) [inline core/server/api/settings.js]: The logic is different. Trying to set the activeTheme via the Settings API is a bad request, nothing to do with permissions.  That's why it happens first.

The other if statement is to do with if you do or don't have permission - and permits the changes for internal requests.

Setting activeTheme this way must never be allowed.

Comment (ErisDS) [inline core/server/api/settings.js]: I do appreciate the desire to have less code - maybe we could also change it so that core settings can only be changed via the model, never the API, but this seems outside the scope of this PR?

File: core/server/themes/validate.js (hunk 1 above)
Comment (ErisDS) [inline core/server/themes/validate.js]: Ah yes meant to change that will do it now 👍

File: core/server/api/themes.js [hunk 3]
```diff
@@ -26,23 +23,41 @@ var debug = require('debug')('ghost:api:themes'),
...
-        });
+                return themeUtils.validate.check(loadedTheme);
+            })
+            .then(function haveValidTheme(_checkedTheme) {
+                checkedTheme = _checkedTheme;
+                // We use the model, not the API here, as we don't want to trigger permissions
+                return settingsModel.edit(newSettings, options);
+            })
+            .then(function hasEditedSetting() {
+                // @TODO actually do things to activate the theme, other than just the setting?
```
Comment (kirrg001) [inline core/server/api/themes.js]: What else need's to be done here except of setting the active theme in the model?

Comment (ErisDS) [inline core/server/api/themes.js]: This is a foreshadowing of my next PR 😉  in which I properly introduce a concept of "activating" a theme.

Atm we have 3 concepts:

1. on boot, sort of activate a theme by doing some loading
2. on activate, sort of activate a theme by changing a setting
3. on upload, sort of activate a theme by overriding and having a [dirty hack](https://github.com/TryGhost/Ghost/blob/master/core/server/api/index.js#L69).

In all 3 cases, the next request magically "fixes" the state here: https://github.com/TryGhost/Ghost/blob/master/core/server/middleware/theme-handler.js

All of which means in the current setup I cannot implement any of the features we need properly without hacks.

Next PR will introduce a better concept of both "activating" and then "mounting" a theme, from which the new features become simple changes.

**Comment by ErisDS** [comment]
Have added an extra commit to move gscan

**Review by kirrg001 — APPROVED** [review]
LGTM 😎

Tested the basics via admin, all works as expected. Found one UI thing with activating themes, but kevin will sort this out as part of #8127

If we find any edge cases, we can fix them in the beta phase 👍
